clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -184,6 +184,18 @@ static bool IsSyclMathFunc(unsigned BuiltinID) {
   return true;
 }
 
+static bool isKnownGoodDecl(const Decl *D) {
+  if (const FunctionDecl *FD = dyn_cast<FunctionDecl>(D)) {
+    const IdentifierInfo *II = FD->getIdentifier();
+    if (II &&
+        FD->getDeclContext()->isTranslationUnit() &&
```

Comment:
This check is incorrect, Consider:
extern "C++" { __spirv_ocl_printf(whatever); }

Or :
extern "C" { extern "C++" { __spirv_ocl_printf(whatever);}}

Also, what if the declaration is a friend declaration?

Finally, make sure !FD->isDefined();

Reply:
> Also, what if the declaration is a friend declaration?

Can you give an example where __spirv_ocl_printf is declared as a friend and should not be accepted?

Reply:
I'm saying it SHOULD be accepted:

class S {
   friend void __spirv_ocl_printf(whatever);
};

Is the declcontext of that functiondecl the class?

Reply:
> I'm saying it SHOULD be accepted:
> 
> class S {
> friend void __spirv_ocl_printf(whatever);
> };
> 
> Is the declcontext of that functiondecl the class?

Ok, for this:
```
class A {
  friend int __spirv_ocl_printf(const char *__format, ...);
};
int main() {
  kernel_single_task<class fake_kernel>([]() { __spirv_ocl_printf("Hello world! %d%d\n", 4, 2); });
}
```
I get an error `use of undeclared identifier '__spirv_ocl_printf'`.
If I also make a proper global declaration of __spirv_ocl_printf, then it works fine (no diagnostic).

Reply:
Yeah, it isn't in scope in that case.  Try:

struct A {
  friend int __spirv_ocl_printf(const char *__format, ...);
};
int __spirv_ocl_printf(const char *__format, ...);
int main() {
  kernel_single_task<class fake_kernel>([]() { __spirv_ocl_printf("Hello world! %d%d\n", 4, 2); });
}

The 'canonical' decl of the function is likely going to be the friend version I think.

Reply:
I added more checkes and extended the test to cover these cases.